packages/yew/src/html/component/lifecycle.rs

```diff
@@ -125,24 +126,25 @@ impl<COMP: Component> Runnable for ComponentRunnable<COMP> {
 
                     let should_render = match event {
                         UpdateEvent::First => true,
-                        UpdateEvent::Message(message) => state.component.update(message),
+                        UpdateEvent::Message(message) => {
+                            state.component.update(&state.scope, message)
+                        }
                         UpdateEvent::MessageBatch(messages) => {
-                            let component = &mut state.component;
-                            messages
-                                .into_iter()
-                                .fold(false, |acc, msg| component.update(msg) || acc)
+                            messages.into_iter().fold(false, |acc, msg| {
+                                state.component.update(&state.scope, msg) || acc
+                            })
                         }
                         UpdateEvent::Properties(props, node_ref, next_sibling) => {
                             // When components are updated, a new node ref could have been passed in
                             state.node_ref = node_ref;
                             // When components are updated, their siblings were likely also updated
                             state.next_sibling = next_sibling;
-                            state.component.change(props)
+                            state.component.changed(&state.scope, Rc::clone(&props))
```

Comment:
Cloning and `Rc` is actually not that cheap. Why not pass a reference ton an `Rc`? This also avoids useless work, if changed simply returns `false`.

Reply:
The user would have to clone it themselves then. The component can't store a reference to props (props need to be stored to be accessed in `view`)

Reply:
>The user would have to clone it themselves then.

A fair trade off, I find. Not every component would need to store it, so you are paying for a both an increment and decrement of the counter with no gain in these cases (likely the majority case).

Reply:
I'm not sure why a component would take a prop but never use it. In order for a prop to be used, it must be stored.

Reply:
On another note, why store the props at all on the component? That means the user has to:
- Add a field to their component
- Init the field in `create()`

Personally, I quite like the Actix's approach of passing a context reference to all user Component calls. You can pass a mutable or immutable reference to differentiate between what `view()` and `update()` or `change()` can call.

Reply:
I believe the best solution is to pass make Context contain both `Scope` and props. But then destroy interferes as it doesn't get the props.

Reply:
I made it so that context contains everything and is passed everywhere. Take a look.